### DISCUSSION TITLE
How do I limit player's movements?

### DISCUSSION THREAD

**Discussion by Mir-Ali-1** [bjornbytes/lovr#663] [General]
I am developing a VR game but the player is able to go through walls and fly.

**Comment by Trixelized** [comment]
The desktop WASD and mouse controls just exist to simulate where the player's head would be if there is nothing to actually track. If you want to disable this for desktop, you can change `t.modules.headset` to `false` in [lovr.conf](https://lovr.org/docs/lovr.conf). You can check if the current driver is desktop with [lovr.headset.getDriver](https://lovr.org/docs/lovr.headset.getDriver).

**Comment by jmiskovic** [comment]
There are at least two approaches. One would be to ignore the player's head moving through walls, but design the experience so that player cannot really take advantage of it - for example using really thick walls and restricting locomotion to teleportation-only. You could also teleport player back if they go too far into the wall. This approach can be done using the physics engine and the colliders (or at least sensor shapes) to detect when collision with wall happens. Details would depend on your overall concept and how the locomotion fits inside it.

The second approach is to actively block the head from moving into the wall. You have to be careful thought, because you usually want to keep 1:1 relation between real world head movement and VR camera movement, or some people might experience disorientation or nausea otherwise. Gorilla Tag uses this approach, both the head and torso cannot ever penetrate the wall. The basic gist is that you use a sphere collider to represent the head, and fixate the camera on this sphere. This gets quite tricky because the sphere position and orientation is controlled from three sources: the head movement, the physics collision response, and the VR locomotion system.

This is actually not an easy problem. It's good to keep it in mind when designing the VR environment.
